components/connector/internal/certificates/loader.go

```diff
@@ -16,42 +18,94 @@ type Loader interface {
 
 type certLoader struct {
 	certificatesCache           Cache
+	readinessCh                 chan<- struct{}
+	apiServerClient             discovery.ServerVersionInterface
 	repository                  secrets.Repository
 	caSecretName                types.NamespacedName
 	rootCACertificateSecretName types.NamespacedName
+	exitCh                      <-chan struct{}
 }
 
 func NewCertificateLoader(certificatesCache Cache,
 	repository secrets.Repository,
 	caSecretName types.NamespacedName,
-	rootCACertificateSecretName types.NamespacedName) Loader {
+	rootCACertificateSecretName types.NamespacedName,
+	readinessCh chan<- struct{}, apiServerClient discovery.ServerVersionInterface,
+	exitCh <-chan struct{}) Loader {
 	return &certLoader{
 		certificatesCache:           certificatesCache,
+		readinessCh:                 readinessCh,
+		apiServerClient:             apiServerClient,
 		repository:                  repository,
 		caSecretName:                caSecretName,
 		rootCACertificateSecretName: rootCACertificateSecretName,
+		exitCh:                      exitCh,
 	}
 }
 
 func (cl *certLoader) Run() {
+	notificationCh := make(chan struct{}, 1)
+	shouldNotify := true
+	go cl.notifyOnConnection(notificationCh)
+
 	for {
-		if cl.caSecretName.Name != "" {
-			cl.loadSecretToCache(cl.caSecretName)
+		allLoaded := cl.loadSecretsToCache()
+		if allLoaded && shouldNotify {
+			shouldNotify = false
+			log.Info("All needed secrets are loaded, notifying readiness")
+			cl.readinessCh <- struct{}{}
```

Comment:
The idea of this readiness handler was to let the application bootstrap flawlessly. That way liveness probe could pass once the application is running (the right configurations were set). On the other hand readiness handler will indicate whether the pod is ready to serve(certs are loaded). If we decide to go with the approach where we wait on bootstrap to load cert, we could find ourself in situation where kubelet restarts the pod endlessly - istio sidecar proxy is not being initialized for the time the liveness probes are made and connector won't be able to fetch the certs. I guess we could increase the initialDelaySeconds of liveness probe for the connector but how could we be sure that this increased time will fit in the initialization time needed for the Envoy proxy(could vary on different environments)? So with that taken into account I decided to introduce this readiness handler and let the application bootstrap flawlessly. I agree that the code could be refactored - new method on certLoader, let's say WaitForCerts could be added, and it will be called  before the actual regular update of the secrets. (shouldNotify will be removed). 

I agree that the suggested solution is more clear and easier to implement and maintain but I'm not sure how we could make sure we won't get trapped in endless loop. Do you have any suggestions on that?